### PR TITLE
zfs: deprecate latestCompatibleLinuxPackages

### PR DESCRIPTION
**Pull request by adamcstephens**
## Description of changes

Using zfs.latestCompatibleLinuxPackages can result in downgrades to the kernel on a system, potentially causing breakage.

This breakage may not be apparent during build and switch, but only after attempting to reboot into the updated generation. 

By forcing users to explicitly manage their kernel version, we can ensure that the breakage will be apparent at build time instead.

Here is but one example of breakage with this option: https://github.com/NixOS/nixpkgs/pull/340382#issuecomment-2348994525

Todo
- [x] Update release notes once agreed on strategy

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by adamcstephens** [comment]
Alternatively, we can just throw and remove the option now. :)

**Review by amarshall — COMMENTED** [review]
Yea I’ve been wanting to do this. It makes sense. The caveat here is that we currently use it for our own tests, so we need to do something there. I have some wip stuff to make the tests test more Kernel versions, and also a separate wip to test our Kernel compat spec against the `META` in the ZFS source. Might be a bit on those, though.

Ofc, none of that blocks this, as it’s just a warning for now, and I agree we should at least mark our intent to remove it.

Not sure if this also warrants an entry in the release notes.

File: pkgs/os-specific/linux/zfs/generic.nix
Comment (adamcstephens) [reply pkgs/os-specific/linux/zfs/generic.nix:209]: Latest LTS is usually the default, so for people who can use that they don't need to specify kernel at all. I'll update the wording.

**Review by endocrimes — CHANGES_REQUESTED** [review]
We should not land this change - the root of kernel downgrades comes moreso from our _aggressive_ removal of kernel releases as they drop out of support, while tools like ZFS lag slightly behind. Keeping those releases until ZFS catches up would likely be a better outcome for most people.

If you need a "more recent than LTS, but still ZFS compatible" kernel, this is the best option we have (and most people don't need the _bleeding_ edge for core device support - if they do, they should be managing this themselves, but those folks shouldn't make managing versions harder for everyone else). We probably do need better documentation for it though.

**Comment by adamcstephens** [comment]
> Keeping those releases until ZFS catches up would likely be a better outcome for most people.

Feel free to try again to convince the kernel team, but my understanding is this is a non-starter. That's the kernel team's prerogative. 

> If you need a "more recent than LTS, but still ZFS compatible" kernel, this is the best option we have

I'd argue that explicitly setting the kernel version is a better option. It avoids hidden downgrades and ensures that users are aware of the impact of changing the kernel. For most users they shouldn't need to touch the kernel version at all, as zfs tends to have good support for the longterm kernels. For everybody else, I'm of the opinion that they should explicitly opt in to specific kernel versions

> We probably do need better documentation for it though.

I'm not convinced that any amount of documentation will fix the fundamental problem that this option will silently downgrade kernels, exposing unaware users to potential breakage. This is a bad behavior that we on the zfs maintenance team (amarshall and I) would rather not support.

Personally I only run zfs 2.2 and on the stable/longterm kernel.

**Comment by adamcstephens** [comment]
This was shared with me by a community member as an alternative for those who *just want latest*. It likely has its problems at times, but we could add it to the wiki with a note about such.

```nix
  nixpkgs.overlays = [
    (self: super: {
      zfs = super.zfs.overrideAttrs ({ meta ? { }, ... }: {
        meta = meta // { broken = false; };
      });

      linuxPackages_latest = super.linuxPackages_latest.extend (self: super: {
        zfs = super.zfs.overrideAttrs ({ meta ? { }, ... }: {
          meta = meta // { broken = false; };
        });
      });
    })
  ];
```

**Comment by cafkafk** [comment]
> I think we’re also not super clear on the risk of using `zfs_unstable`, either.

Naming it zfs_master instead would be much clearer IMO, specially coming from a NixOS perspective of master meaning the actual latest commit from upstream (unlike e.g. nixos-unstable which does actually have a fairly decent amount of guarantees)

**Comment by adamcstephens** [comment]
> Naming it zfs_master instead would be much clearer IMO

Actually, it wouldn't. The zfs_unstable package is either pointed at the latest release, or the staging branch for the next release, whichever is newest. It should not be pulling from master, as I understand it upstream only runs their test suite against releases and staging branches.

**Comment by cafkafk** [comment]
so actually, it should be called `zfs_staging` (and perhaps have a version set to e.g. `2.2.7-staging`)? seems like that alone would communicate a lot of important information to the user

**Comment by MarcelCoding** [comment]
How about renaming the option into something like `latestCompatibleLtsLinuxPackages` and only let this be a lts release, don't update to non-lts releases? Then zfs users (like) me had a way to manage their kernel version to be compatible with zfs, don't have to deal with downgraded and also don't need to pin the kernel. I am strongly against removing this option of managing the kernel version without adding an alternative (e.g. the one I've proposed.)

**Comment by adamcstephens** [comment]
> How about renaming the option into something like `latestCompatibleLtsLinuxPackages` and only let this be a lts release, don't update to non-lts releases?

This option would be nothing like the one I'm suggesting we remove, and also would be completely unnecessary 90+% of the time. If you don't want to manage your kernel version, and are fine with the LTS kernel, you can just use the default kernel and zfs versions.

**Comment by adamcstephens** [comment]
To give a more detailed breakdown on how an LTS option would work.

2022/12/12 - kernel 6.1 added to nixpkgs
2023/01/21 - zfs 2.1.8 updated in nixpkgs, adding kernel 6.1 support
2023/02/10 - kernel 6.1 becomes default in nixpkgs

2023/10/30 - kernel 6.6 released and added to nixpkgs
2023/11/26 - zfs 2.2.1 updated in nixpkgs, adding kernel 6.6 support
2024/02/02 - kernel 6.6 becomes default in nixpkgs

This means that during the 6.1 cycle, an option pinning to latest LTS (which does not appear to be actually codified would likely require changes to the kernel), would get you:

20 days on 6.1 before it became the default
2 months on 6.6 before it became the default

For the rest of the time, an option for tracking latest LTS supported by zfs would provide no benefit over the defaults.

---

For non-LTS users, the current option is definitely much more aggressive and provides larger coverage, but it comes with the downsides listed above.

**Comment by MarcelCoding** [comment]
Are there any guarantees that the zfs updates get into nixpkgs before the default kernel changes or is this just luck and just happened to be in the correct order in the recent past?

**Comment by MarcelCoding** [comment]
ok, that's something that I could live with

File: pkgs/os-specific/linux/zfs/generic.nix
Comment (Atemu) [reply pkgs/os-specific/linux/zfs/generic.nix:209]: Could you remove all the complex selection machinery and just make it point to the default `linuxPackages`?

Comment (Atemu) [reply pkgs/os-specific/linux/zfs/generic.nix:209]: Could you remove all the complex selection machinery and just make it point to the default `linuxPackages`?

Comment (Atemu) [reply pkgs/os-specific/linux/zfs/generic.nix:209]: Nit:

```suggestion
        (lib.warn "zfs.latestCompatibleLinuxPackages is deprecated and will be removed prior to 25.05. If using the stable LTS kernel (default `linuxPackages`) is not possible then you should explicitly pin a specific kernel release. For example, `boot.kernelPackages = pkgs.linuxPackages_6_6`. Please be aware that non-LTS kernels are likely to go EOL before ZFS supports the latest supported non-LTS release.")
```

**Review by Atemu — APPROVED** [review]
This should have never been a thing if I'm honest and the automatic selection iteration that we have now is even worse. Let's rip off the band-aid.

Could you just yank the code completely and make this an alias for the default `linuxPackages`?

**Comment by adamcstephens** [comment]
Updated to just alias to default kernel with warning and add to release notes. We can remove post 24.11 branch off.

**Review by Atemu — APPROVED** [review]
Will merge in a few days if nobody objects.

**Comment by SuperSandro2000** [comment]
> I'm not sure if there are any guarantees, but if the LTS kernel was updated prior to zfs supporting it then your configuration would fail at eval time.

We could avoid that by going back to what we had before with one catch: instead of naming it `latestCompatibleLinuxPackages` we name it `compatibleLinuxPackages` which points by default to linuxPackages but in case that would be bumped without being compatible with the current zfs version we could temporarily roll back to an older kernel without users having to manually adjust their kernel version and then maybe forgetting that they can go off the older kernel again.

**Comment by cafkafk** [comment]
Honestly should call it something like `zfsLatestCompatibleKernelRollercoaster`, at least that's very explicit

**Review by amarshall — APPROVED** [review]

**Comment by adamcstephens** [comment]
Thanks for trying to make things better. :)

**Comment by RaitoBezarius** [comment]
Hey guys, I will probably soon deprecate ZFS 2.1.x because I will not maintain it anymore in nixpkgs. So no pressure on 2.1.x series.

**Comment by Mic92** [comment]
Can I convince you to name the attribute `zfsLatestCompatibleKernelRollercoaster` instead, basically the trainings-wheel free version?
I use it to build installer images and I want the latest kernel if possible to have better hardware support.

**Comment by cafkafk** [comment]
> Can I convince you to name the attribute `zfsLatestCompatibleKernelRollercoaster` instead, basically the trainings-wheel free version? I use it to build installer images and I want the latest kernel if possible to have better hardware support. Also newer kernel generally have less known security vulnerabilities because not everything gets backported.

I actually, unironically, would love to have this available.

**Comment by Atemu** [comment]
Note that you could just copy the mechanism to select the latest compatible linuxPackages to your own config if you really need it.

It's not something we should have ever had in Nixpkgs though IMHO as its quite fragile and uses tryEval on a bunch of packages. I'd strongly request anyone attempting to include such code into Nixpkgs to reconsider.

The alternative of manually specifying versions always meant increased maintenance overhead for the ZFS maintainers and I think it's clear that they don't want to keep doing it but I can't speak for them.

**Comment by Mic92** [comment]
> The alternative of manually specifying versions always meant increased maintenance overhead for the ZFS maintainers and I think it's clear that they don't want to keep doing it but I can't speak for them.

ZFS maintainer still need to specify compatible version bounds because zfs is build for multiple versions. There is no maintenance increase coming from this. But I guess I am tired of having these things change all the time anyway. Once copied in I probably never need to touch it again. Everyone that needs it can pull it from srvos.nixosModules.mixins-latest-zfs-kernel: https://github.com/nix-community/srvos/pull/516

**Comment by Mic92** [comment]
Latest lts doesn't has bcachefs support, which is important for installers. Also most often zfsUnstable supports versions newer than lts, so the majority of time it will be newer than the latest lts version.

**Comment by Mic92** [comment]
I also support a research cluster where we often switch to very current zfs revisions to support newer kernels than zfsUnstable to unlock other very new kernel features. The machines itself have backups, so this is not a concern, but snapshots are nice to have anyway.

**Comment by Atemu** [comment]
> ZFS maintainer still need to specify compatible version bounds because zfs is build for multiple kernel versions. There is no maintenance increase coming from this.

What I meant was the ZFS maintainers manually needing to keep track of which kernel is the latest that is still supported by ZFS which would be the alternative to the flawed and fragile automatic selection mechanism.

> Latest lts doesn't has bcachefs support, which is important for installers.

This does not answer the question of how this changes anything about the problem of the kernel version very likely ending up at LTS at one point or another anyways. Wouldn't your installer just suddenly break because bcachefs would suddenly stop working in such a case?

ZFS+bcachefs support in the same installer just simply cannot be supported in a safe and sustainable manner at this time. That and bcachefs being new and experimental enough warrant a separate installer IMHO. (Perhaps even using Kent's tree with the newest fixes as not everything makes it into mainline in a timely manner AFAIA.)

> I also support a research cluster where we often switch to very current zfs revisions to support newer kernels than zfsUnstable to unlock other very new kernel features. The machines itself have backups, so this is not a concern, but snapshots are nice to have anyway.

I don't see how that's relevant to this case as you'd have to pin the kernel version yourself anyways.

(OT but I wouldn't trust any backups made from a potentially broken filesystem implementation. I'd rather look into porting the new shiny features to an LTS kernel rather than porting ZFS to the newest kernel.)

**Comment by ElvishJerricco** [comment]
I do wish this PR had at least added the latest compatible Linux *version number* as a passthru attribute. This doesn't have to be a `linuxPackages` that nixpkgs actually packages; it's just useful to expose the kernel version range that ZFS is compatible with. Then, it would be trivial for out-of-tree uses to re-implement this logic (or improve upon it with broader kernel support ranges) without the nasty `tryEval` / `meta.broken` mechanism. Is that something we could do?

**Pull request merged**
